### PR TITLE
feat(playwright-test): introduce first-party docker integration

### PR DESCRIPTION
**Pull request by aslushnikov**
This patch adds an experimental capability to run Playwright Test
browsers in Docker container.

To do so, run tests with the `PW_GRID=docker npx playwright test`
command.

### CODE REVIEW

**Comment by pavelfeldman** [comment]
Let's land the lego blocks for the generic grid first?

File: src/grid/gridServer.ts [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+  }
+
+  port(): number {
+    return this._server.port();
+  }
+
+  async stop() {
+    for (const agent of this._agents.values())
+      agent.close();
+    this._agents.clear();
```
Comment (aslushnikov) [inline src/grid/gridServer.ts:271]: assert that agents are empty

File: src/test/runner.ts [hunk 2]
```diff
@@ -181,6 +182,14 @@ export class Runner {
     let globalSetupResult: any;
     if (config.globalSetup)
       globalSetupResult = await (await this._loader.loadGlobalHook(config.globalSetup, 'globalSetup'))(this._loader.fullConfig());
+    let gridServer: GridServer | undefined;
```
Comment (aslushnikov) [inline src/test/runner.ts:185]: explicit grid start instead of this

File: src/grid/gridServer.ts [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+class GridAgent extends EventEmitter {
+  private _capacity: number;
+  readonly agentId = createGuid();
+  private _ws: WebSocket | undefined;
+  readonly _workers = new Map<string, GridWorker>();
+  private _status: AgentStatus = 'none';
+  private _workersWaitingForAgentConnected: GridWorker[] = [];
+  private _retireTimeout: NodeJS.Timeout | undefined;
+  private _log: debug.Debugger;
+  private _tenantId: string;
```
Comment (aslushnikov) [inline src/grid/gridServer.ts:74]: s/tenantId/authToken

File: src/grid/gridServer.ts [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+        return !!agentId && this._agents.has(agentId);
+      }
+
+      return false;
+    };
+
+    this._wsServer.on('connection', async (ws, request) => {
+      if (request.url?.startsWith('/claimWorker')) {
+        const params = new URL('http://localhost/' + request.url).searchParams;
+        const tenantId = params.get('tenantId');
```
Comment (aslushnikov) [inline src/grid/gridServer.ts:191]: have a single `--factory=`

File: src/grid/gridServer.ts [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+      response.end(this._state());
+      return true;
+    });
+
+    this._wsServer = this._server.createWebSocketServer();
+
+    this._wsServer.shouldHandle = request => {
+      this._log(request.url);
+      if (request.url!.startsWith('/claimWorker')) {
+        // shouldHandle claims it accepts promise, except it doesn't.
```
Comment (aslushnikov) [inline src/grid/gridServer.ts:175]: handle autherization here

File: src/grid/gridServer.ts [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+  }
+}
+
+export class GridServerClient {
+  private _ws: WebSocket;
+  private _playwright: Playwright;
+
+  static async create(gridURL: string) {
+    const ws = new WebSocket(`${gridURL}/claimWorker?tenantId=docker`);
+    await new Promise(f => ws.once('message', f));
```
Comment (aslushnikov) [inline src/grid/gridServer.ts:282]: race with close

File: src/grid/gridServer.ts [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+      if (request.url?.startsWith('/claimWorker')) {
+        const params = new URL('http://localhost/' + request.url).searchParams;
+        const tenantId = params.get('tenantId');
+        if (!tenantId) {
+          ws.close();
+          return;
+        }
+        const agent = [...this._agents.values()].find(w => w.canCreateWorker(tenantId)) || this._createAgent(tenantId);
+        if (!agent) {
+          ws.close();
```
Comment (aslushnikov) [inline src/grid/gridServer.ts:198]: close with custom error

File: src/grid/gridServer.ts [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+
+      return false;
+    };
+
+    this._wsServer.on('connection', async (ws, request) => {
+      if (request.url?.startsWith('/claimWorker')) {
+        const params = new URL('http://localhost/' + request.url).searchParams;
+        const tenantId = params.get('tenantId');
+        if (!tenantId) {
+          ws.close();
```
Comment (aslushnikov) [inline src/grid/gridServer.ts:193]: close with custom reason

**Review by aslushnikov — COMMENTED** [review]
Pavel review comments

**Comment by aslushnikov** [comment]
Closing in favor of #8941

**Pull request closed without merging**